### PR TITLE
GeometricSurface Interface added + Utilities

### PR DESCRIPTION
**Pull request by jklamer**
Adding in the GeometricSurface Interface to handle geometric surface interactions (namely Polygon and Multipolygon). Then used the interface to expand the usability of AtlasEntities and clean up the AtlasEntityPolygonFilter
Things to note:

- ```AtlasEntity::intersects(Polygon)``` ->  ```AtlaEntity::intersects(GeometricSurface)``` 

- I implemented ```MultiPolygon::fullyGeometricallyEncloses(MultiPolygon)```, ~~~```MultiPolygon::fullyGeometricallyEncloses(Segment)```~~~ , and ```MultiPolygon::surfaceOnSphere()```
- Backwards Compatible
- Some utility added in QuadTree

Some questions I have:
- Should I get ride of ```IntersectionPolicy ```  interface and replace it with a ```BiPredicate<GeometricSurface,AtlasEntity>``` ?
- Are there more functions to add to the Interface ?
- Should I allow overlapping GeometricSurfaces in AtlasEntityPolygonFilter ?

### CODE REVIEW

**Comment by jklamer** [comment]
Also PR number #100 !!! celebrate.gif

File: src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java [hunk 1]
```diff
@@ -238,6 +240,55 @@ public boolean fullyGeometricallyEncloses(final PolyLine polyLine)
...
+            if (!cleared)
+            {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public boolean fullyGeometricallyEncloses(final Segment segment)
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java]: Something we discussed offline: could we potentially leverage this method when looking at `fullyGeometricallyEncloses(final PolyLine polyline)`? That might need more testing.

File: src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java [hunk 2]
```diff
@@ -238,6 +240,55 @@ public boolean fullyGeometricallyEncloses(final PolyLine polyLine)
...
+            }
+            if (!cleared)
+            {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java]: To be consistent, can we add a javadoc here as well.

File: src/test/java/org/openstreetmap/atlas/utilities/filters/AtlasEntityPolygonsFilterTest.java [hunk 3]
```diff
@@ -99,6 +100,37 @@ public boolean polygonEntityIntersecting(final Polygon polygon, final AtlasEntit
                 return false;
             }
         }
+
+        // note this is the only one now used by the filter
```
Comment (MikeGost) [inline src/test/java/org/openstreetmap/atlas/utilities/filters/AtlasEntityPolygonsFilterTest.java]: What does this comment mean?

Comment (jklamer) [reply src/test/java/org/openstreetmap/atlas/utilities/filters/AtlasEntityPolygonsFilterTest.java]: The geometricSrufaceEntityIntersecting is the only function that is being used now. Which relates to my overall PR question of should I get rid of the Intersection Policy Interface and replace it with a customizable BiPredicate of AtlasEntity and GeometricSurface that I cast to serializable?

File: src/main/java/org/openstreetmap/atlas/geography/atlas/items/Area.java [hunk 4]
```diff
@@ -62,9 +63,9 @@ public ItemType getType()
     }
 
     @Override
-    public boolean intersects(final Polygon polygon)
+    public boolean intersects(final GeometricSurface geometricSurface)
```
[the remark this answers is not in the document]
Comment (jklamer) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/items/Area.java]: sure, we already have a class called surface so I just avoided it first go through.

File: src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java (hunk 1 above)
Comment (jklamer) [reply src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java]: Did some performance testing and this is not as performant so I am going to remove and remove from the interface.

File: src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java [hunk 5]
```diff
@@ -30,14 +31,13 @@
  *
  * @author matthieun
  */
-public class MultiPolygon implements Iterable<Polygon>, Located, Serializable
+public class MultiPolygon implements Iterable<Polygon>, GeometricSurface, Located, Serializable
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java]: You can remove the `Located` here!

File: src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java [hunk 6]
```diff
@@ -219,23 +220,50 @@ public boolean fullyGeometricallyEncloses(final Location location)
...
+     * {@link MultiPolygon}
+     * 
+     * @param that
+     *            the provided {@link MultiPolygon} to test
+     * @return true if the conditions are met, false otherwise
+     */
+    @Override
+    public boolean fullyGeometricallyEncloses(final MultiPolygon that)
+    {
+        final QuadTree<Polygon> thisOuters = QuadTree.forLocated(this.outers());
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java]: `QuadTree` is usually a good choice for Point objects. For anything with a shape, I would recommend using an `RTree` instead.

Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java]: Did you test the performance on this one out of curiosity?

File: src/main/java/org/openstreetmap/atlas/geography/Polygon.java [hunk 7]
```diff
@@ -190,6 +191,12 @@ public boolean fullyGeometricallyEncloses(final Location location)
         }
     }
 
+    @Override
+    public boolean fullyGeometricallyEncloses(final MultiPolygon multiPolygon)
+    {
+        return multiPolygon.outers().stream().allMatch(this::fullyGeometricallyEncloses);
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/Polygon.java]: What if this polygon entirely falls within an inner polygon of the multipolygon passed in? I believe this is not a covered use case here? If not, mind adding a test for that specific corner case?

File: src/main/java/org/openstreetmap/atlas/geography/index/QuadTree.java [hunk 8]
```diff
@@ -33,6 +36,21 @@
     private final Quadtree tree = new Quadtree();
     private Rectangle bound;
 
+    public static <K extends Located> QuadTree<K> forLocated(final Collection<K> locatedCollection)
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/index/QuadTree.java]: If you use only `forEach` how about making it more generic and using `Iterable` instead of `Collection`?

File: src/main/java/org/openstreetmap/atlas/geography/index/QuadTree.java [hunk 9]
```diff
@@ -33,6 +36,21 @@
...
+    {
+        final QuadTree<K> toReturn = new QuadTree<>();
+        locatedCollection.forEach(located -> toReturn.add(located.bounds(), located));
+        return toReturn;
+    }
+
+    public static <K> QuadTree<K> forCollection(final Collection<K> collection,
```
Comment (matthieun) [inline src/main/java/org/openstreetmap/atlas/geography/index/QuadTree.java]: Same comment here.

File: src/test/java/org/openstreetmap/atlas/utilities/filters/AtlasEntityPolygonsFilterTest.java (hunk 3 above)
Comment (MikeGost) [reply src/test/java/org/openstreetmap/atlas/utilities/filters/AtlasEntityPolygonsFilterTest.java]: Good question. Personally I think having a BiPredicate of something like that can get fairly messy and abstracting it away as a policy makes more sense. Open to other ideas though :)

File: src/main/java/org/openstreetmap/atlas/geography/Polygon.java (hunk 7 above)
Comment (jklamer) [reply src/main/java/org/openstreetmap/atlas/geography/Polygon.java]: For a polygon to geometrically enclose a multi-polygon, it must geometrically enclose all the outer polygons of the multi polygon. If the polygon lies within the inner of the passed in multi-polygon then, the polygon does not geometrically enclose the passed in multi-polygon.

File: src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java (hunk 6 above)
Comment (jklamer) [reply src/main/java/org/openstreetmap/atlas/geography/MultiPolygon.java]: I didn't take any time statistics. Are there two different options you would like me to compare the performance of?

File: src/main/java/org/openstreetmap/atlas/geography/Polygon.java (hunk 7 above)
Comment (jklamer) [reply src/main/java/org/openstreetmap/atlas/geography/Polygon.java]: I added a test case to display the interaction I believe to be desirable based on what I understand of the method.
